### PR TITLE
[IE][VPU]: Completion of Configuration options refactoring in VPU plugins

### PR DESCRIPTION
**Pull request by ZlobinGM**
# Task

#-41782

# Description 

  Completion of configuration options refactoring in VPU plugins, that started in PR: [[IE][VPU]: Configuration options in VPU plugins refactoring](https://github.com/openvinotoolkit/openvino/pull/3211)

- [x] compilerLogFilePath - private, remove
- [x] copyOptimization - private, keep 
- [x] customLayers - public, keep 
- [x] detectBatch - private, keep 
- [x] deviceConnectTimeout - private, keep 
- [x] deviceName - public DEVICE_ID, keep 
- [x] disableConvertStages - private, deprecated, keep (but plan to remove it in future)
- [x] disableReorder - private, deprecated, keep (but plan to remove it in future)
- [x] dumpAllPasses - private, keep 
- [x] dumpInternalGraphDirectory - private, keep, rename to dumpAllPassesDirectory 
- [x] dumpInternalGraphFileName - private, keep 
- [x] enableMemoryTypesAnnotation - private, keep 
- [x] enablePermuteMerging - private, keep 
- [x] enableReplaceWithReduceMean - private, remove 
- [x] enableReplWithSCRelu - private, keep 
- [x] enableTensorIteratorUnrolling - private, keep (but plan to remove it in future) 
- [x] enableWeightsAnalysis - private, keep 
- [x] exclusiveAsyncRequests - public, keep 
- [x] forceDeprecatedCnnConversion - private, remove 
- [x] forcePureTensorIterator - private, keep (but plan to remove it in future) 
- [x] forceReset - public, deprecated, keep but ignore 
- [x] hwBlackList - private, keep 
- [x] hwDilation - private, keep 
- [x] hwExtraSplit - private, keep 
- [x] hwOptimization - public, keep 
- [x] hwWhiteList - private, keep 
- [x] ignoreUnknownLayers - private, keep 
- [x] injectSwOps - private, keep 
- [x] ioStrides - private, keep 
- [x] irWithVpuScalesDir - private, keep (but this option does not work right now) 
- [x] logLevel - public, keep 
- [x] memoryType - public, keep
- [x] mergeHwPoolToConv - private, keep 
- [x] noneLayers - private, keep (fix) 
- [x] numCMXSlices - private, keep 
- [x] numExecutors - public, keep 
- [x] numSHAVEs - private, keep 
- [x] packDataInCmx - private, keep 
- [x] perfCount - public, keep 
- [x] perfReport - private, keep 
- [x] platform - public, deprecated, keep but ignore 
- [x] pluginLogFilePath - private, remove 
- [x] powerConfig - private, keep 
- [x] printReceiveTensorTime - public, keep 
- [x] protocol - public, keep 
- [x] tilingCMXLimitKB - private, keep 
- [x] watchdogInterval - private, keep

### CODE REVIEW

File: inference-engine/include/ie_parameter.hpp [hunk 1]
```diff
@@ -318,6 +353,20 @@ class INFERENCE_ENGINE_API_CLASS(Parameter) {
         bool operator==(const Any& rhs) const override {
             return rhs.is(typeid(T)) && equal<T>(*this, rhs);
         }
+
+        template <class U>
+        typename std::enable_if<!HasOutputStreamOperator<U>::value, void>::type
+        print(std::ostream& stream, const U& object) const {}
```
Comment (pelszkow) [inline inference-engine/include/ie_parameter.hpp:359]: Consider don't mix return type with templates/concepts:
```suggestion
        template <class U,  
                  typename std::enable_if<!HasOutputStreamOperator<U>::value, bool>::type = true>
        void print(std::ostream& stream, const U& object) const {}
```

In our src base `int` is sometimes used:
```suggestion
        template <class U,  
                  typename std::enable_if<!HasOutputStreamOperator<U>::value, int>::type = 0>
        void print(std::ostream& stream, const U& object) const {}
```

File: inference-engine/include/ie_parameter.hpp [hunk 2]
```diff
@@ -273,6 +289,24 @@ class INFERENCE_ENGINE_API_CLASS(Parameter) {
...
+        static auto test(...) -> std::false_type {
+            return {};
+        }
+
+        using type = typename std::is_same<std::true_type, decltype(test<T, U>(nullptr))>::type;
+    };
+
+    template <class T>
+    struct HasOutputStreamOperator : CheckOutputStreamOperator<std::ostream, T>::type {};
+
```
Comment (pelszkow) [inline inference-engine/include/ie_parameter.hpp:309]: This looks like mix of c++98 and c++11 SFINAE
consider this https://godbolt.org/z/KK8exxjxc

File: inference-engine/src/inference_engine/cpp/ie_infer_request.cpp [hunk 3]
```diff
@@ -374,7 +374,7 @@ std::vector<VariableState> InferRequest::QueryState() {
 }
 
 bool InferRequest::operator!() const noexcept {
-    return !_impl || !actual;
+    return !_impl && !actual;
```
Comment (pelszkow) [inline inference-engine/src/inference_engine/cpp/ie_infer_request.cpp:377]: Consider
```suggestion
    return !static_cast<bool>(*this);
```

**Review by AndrewBakalinIntel — APPROVED** [review]
Great job 👍 
Left a minor comment.

File: inference-engine/src/vpu/graph_transformer/src/graph_transformer.cpp [hunk 4]
```diff
@@ -81,39 +89,36 @@ void CompileEnv::init(ncDevicePlatform_t platform, const PluginConfiguration& co
 #endif
 
     if (platform == ncDevicePlatform_t::NC_MYRIAD_2) {
-        g_compileEnv->config.compileConfig().hwOptimization = false;
+        g_compileEnv->config.set(ie::MYRIAD_ENABLE_HW_ACCELERATION, ie::PluginConfigParams::NO);
```
Comment (AntonDudchenko) [inline inference-engine/src/vpu/graph_transformer/src/graph_transformer.cpp:92]: I think you can remove this check at all, since Myriad2 is not supported

**Comment by AntonDudchenko** [comment]
LGTM, but I'm afraid that this PR and https://github.com/openvinotoolkit/openvino/pull/6022 conflict. 
@DariaMityagina FYI

**Comment by AntonDudchenko** [comment]
@ArtemySkrebkov-intel, please take a look

**Review by AntonDudchenko — APPROVED** [review]

File: inference-engine/src/vpu/common/src/configuration/parse_numeric.cpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "vpu/configuration/parse_numeric.hpp"
+
+#include "ie_plugin_config.hpp"
+
+#include "vpu/utils/error.hpp"
+
+namespace vpu {
+
+int parseInt(const std::string& src) {
```
[the remark this answers is not in the document]
Comment (ZlobinGM) [reply inference-engine/src/vpu/common/src/configuration/parse_numeric.cpp]: Agree. Fixed

File: inference-engine/src/vpu/graph_transformer/src/graph_transformer.cpp (hunk 4 above)
Comment (ZlobinGM) [reply inference-engine/src/vpu/graph_transformer/src/graph_transformer.cpp:92]: I think this should be done in a separate PR on the removal of the PLATFORM option.

Comment (DariaMityagina) [reply inference-engine/src/vpu/graph_transformer/src/graph_transformer.cpp:92]: There is such a [PR](https://github.com/openvinotoolkit/openvino/pull/6022).

**Review by ArtemySkrebkov-intel — APPROVED** [review]

**Pull request merged**